### ISSUE TITLE
Feature: implement Kotlin's 'when'-block

### ISSUE THREAD

**Issue by Mwexim** [SkriptLang/Skript#4000]
Before I start this suggestion, I want you to know that I am aware that this is very similar to the countlessly proposed 'switch section' suggestions. I still want to make this feature request, since it is quite different to the current suggestions.

For those who don't know, Kotlin's 'when'-block is basically an enhanced switch section. It allows more customizability, is way easier to use and the syntax is quite simple and straightforward. I actually like it a lot when programming in Kotlin.
*(Note that the examples are ridiculous and do not represent real use-cases)*
```kt
when ("Hello world") {
    "Bye world" -> false       // Compare by value
    "Hello world" -> true
    is Int -> false                 // Compare by using a condition (instanceof)
    else -> false
}
```

There are 4 key points:
1. If the condition is omitted, the values are checked for equality
2. If the condition is present, the input value (`"Hello world"` in this example" will be put before the condition, as if it was pasted before compiling.
3. This block can be used as an expression and as a statement
4. If no input is given, you can put whole conditions as well, like:
```kt
when {
    "orange" is String -> true
    else -> false
}
```
Possible Skript syntax could be:
```vb
when {var}
    is "Hello world", then send it to console
    is "Bye world", then send {var} to console
    is "Good morning world":
        send {var} to console
    is "Oi cunt" or "Nihao":
        send "Something" to console
    + 3 is equal to 5, then send "%it%" to console
    otherwise:
        send "Nothing matched"
```
Note how this example shows a lot of different discussion points:
- Omitting/keeping the semicolon (':') after the section
- Omitting/keeping 'is' for equality checks. For other conditions, the syntax is needed
- Adding 'it' as a reference to the value checked.
- Keeping the cases as sections or allowing inline effects
I will leave these as discussion points, without giving my opinion on them.

**Comment by Mwexim** [comment]
I am planning on adding a similar enhancement to the skript-parser, which already has the basic Java functionality of the switch block. I will wait out discussions on this as I want to gather opinions.

**Comment by nfell2009** [comment]
When to me implies a meaning of a while loop.

**Comment by Itzdlg** [comment]
> When to me implies a meaning of a while loop.

Yeah, Kotlin uses odd names for things, but **when** implies that you are checking for value. type, etc. Switch can be associated with just value, so perhaps a different name?

**Comment by TheLimeGlass** [comment]
fyi SkQuery does this
```vb
when [{checkCount} is set]->[size of all players is greater than or equal to 100]:
    trigger
```

**Comment by ShaneBeee** [comment]
I edited your message, please refrain from using profanity in issues, especially when they do not pertain to the case at hand.

**Comment by Moderocky** [comment]
In English, using “when” would imply waiting for something to happen. The average Skript user doesn’t have programming familiarity and so could be confused by this and assume it’s some kind of wait-for clause.

I would suggest just using “if” as the key word for this, to avoid confusion. I know “if” is already used for if-statements, but this is basically glorified if-statement sugar anyway, so I don’t see why it shouldn’t be used.

**Comment by FranKusmiruk** [comment]
> In English, using “when” would imply waiting for something to happen. The average Skript user doesn’t have programming familiarity and so could be confused by this and assume it’s some kind of wait-for clause.
> 
> I would suggest just using “if” as the key word for this, to avoid confusion. I know “if” is already used for if-statements, but this is basically glorified if-statement sugar anyway, so I don’t see why it shouldn’t be used.

I like the idea, we could call the current if statements well, if statements and the new ones "if sections".

**Comment by Mwexim** [comment]
> OP, I edited your original message, please refrain from using profanity in issues, especially when they do not pertain to the case at hand.

I did not recall using swear words. Could you perhaps tell me what you removed?

**Comment by Mwexim** [comment]
> In English, using “when” would imply waiting for something to happen. The average Skript user doesn’t have programming familiarity and so could be confused by this and assume it’s some kind of wait-for clause.
> 
> 
> 
> I would suggest just using “if” as the key word for this, to avoid confusion. I know “if” is already used for if-statements, but this is basically glorified if-statement sugar anyway, so I don’t see why it shouldn’t be used.
> 
> 
> 
> 

This is a very interesting take and I quite like it.

**Comment by TheLimeGlass** [comment]
To be honest when playing around with these, they do not provide any more beneficial functionality than a while loop and if condition. Also is it going to timeout? Or is it going to infinitely run causing a memory leak? Bukkit and Minecraft will have things that won't trigger it, like locations. I don't know how Kotlin's works.

In terms of Skript parser, this can be a good addition. It's more of a feature enchantment over just using a while and condition.

**Comment by FranKusmiruk** [comment]
> To be honest when playing around with these, they do not provide any more beneficial functionality than a while loop and if condition. Also is it going to timeout? Or is it going to infinitely run causing a memory leak? Bukkit and Minecraft will have things that won't trigger it, like locations. I don't know how Kotlin's works.
> 
> In terms of Skript parser, this can be a good addition. It's more of a feature enchantment over just using a while and condition, but I wouldn't add it to Skript Minecraft.

It isn't a loop, it is a suggestion for a switch statement

**Comment by GovindasOM** [comment]
I'd absolutely love this
